vital/types/track_set.h

```diff
@@ -646,17 +647,21 @@ class VITAL_EXPORT track_set
     return impl_->offset_to_frame(offset);
   }
 
-  virtual track_set_sptr clone() const;
+  virtual track_set_sptr clone( clone_type = clone_type::DEEP ) const;
 
   /// Merges the other feature track set into this feature track set.
   /**
-  * \param [in] other  the other feature track set to merge into this one
-  * \param [in] do_not_append_tracks if true, the other tracks are cloned and assigned a new track id.
-  *  if false, if the same track id is found in other and in the current feature track set, then
-  *  track states from other are cloned and appended to this object's track.
+  * \param other The other feature track set to merge into this one.
+  * \param clone_method How to clone track states, if needed.
+  * \param do_not_append_tracks
+  *   If \c true, the other tracks are cloned and assigned a new track id. If
+  *   \c false, if the same track id is found in other and in the current
+  *   track set, then track states from other are cloned and appended to this
+  *   object's track.
   */
   virtual void merge_in_other_track_set(
     track_set_sptr other,
+    clone_type clone_method = clone_type::DEEP,
```

Comment:
I don't think this one should default to `DEEP`.  We merge tracks frequently and then often throw away `other` tracks that were merged in shortly after.  Normally there is no need for this to be a deep copy.

In fact, I'm wondering more now if `SHALLOW` should just be the default for all the clone methods, preserving the previous behavior.  We could then explicitly use `DEEP` for the special cases that really need a deep copy.

Reply:
> In fact, I'm wondering more now if SHALLOW should just be the default for all the clone methods.

That is going to be awkward for code that actually needs a deep copy, since we won't necessarily be able to pass the argument :cry:.

In particular, [SEAL-TK](https://github.com/kitware/seal-tk) has a class `UnsharedPointer` that "converts" a shared pointer to something with value semantics (i.e. making a copy makes a *copy* — necessarily a *deep* copy — rather than incrementing a reference count). This is a generic class that relies on calling `clone()` on the held object. It's also a *generic* reference holder. I'd prefer to not have to specialize it for every class that needs special gymnastics to make a deep copy.

The alternative is to modify *every* `clone()` in KWIVER.

(I'll also point out that we're discussing "safe but slow by default" versus "fast but broken by default".)

Reply:
I would not say shallow copies are broken by default.  The data on feature tracks was designed to be shallow copied.  Switching to deep copies can break code.  I don't recall if it is still the case, but at some point we were using the fact that two track states shared a pointer to the same data to check that these states are related (derived from a common feature).

Speed and memory usage are very important factors in TeleSculptor, so I'm trying to be careful about changes that will impact that by default.

That said, I'm okay with switching to `DEEP` by default in `clone` and then checking what impact it has.  If we are going to leave `DEEP` as the default in the merge function then we need to carefully  examine each cases where this merge function is called to make sure that `DEEP` is really needed.  This function is mostly used to "move" track states from one set of tracks into another.  After cloning, the source tracks are typically short lived.

Reply:
> I would not say shallow copies are broken by default.

By "broken" I specifically mean "unsafe". And they are, unquestionably. I've made a "copy", but it doesn't actually behave as a copy; it shares state with the original. In a multi-threaded environment, this is *very dangerous*.

> at some point we were using the fact that two track states shared a pointer to the same data to check that these states are related

This almost sounds like an argument for CoW. I think it's really unfortunate that KWIVER wasn't written with CoW semantics from the get-go; they are *much* safer and *much* easier to use. However, I have already spent "too much" time on this. Ultimately, I need safe copies of [object] tracks.

>  If we are going to leave `DEEP` as the default in the merge function [...]

I already changed it to `SHALLOW`. In fact the place where SEAL-TK was using this function was already specifying `SHALLOW`, so it's really not worth arguing for `DEEP`. (I'm not actually sure what "should" be the default, but unlike `clone()`, this is not a generic function, so it doesn't really matter.)